agents/rust/mediator/mediation/src/routes/pickup.rs

```diff
@@ -4,66 +4,81 @@ use std::sync::Arc;
 
 use axum::{extract::State, http::StatusCode, Json};
 use log::info;
-
-use crate::{
-    didcomm_types::{
-        pickup_delivery_message_structs::*, PickupDeliveryReqMsg, PickupMsgEnum, PickupStatusMsg,
-        PickupStatusReqMsg, ProblemReportMsg,
+use messages::{
+    decorators::attachment::{Attachment, AttachmentData, AttachmentType},
+    msg_fields::protocols::pickup::{
+        Delivery, DeliveryContent, DeliveryRequestContent, Pickup, Status, StatusContent,
+        StatusDecorators, StatusRequestContent,
     },
-    storage::MediatorPersistence,
 };
+use uuid::Uuid;
+
+use crate::storage::MediatorPersistence;
 
-pub async fn handle_pickup<T: MediatorPersistence>(
+pub async fn handle_pickup_authenticated<T: MediatorPersistence>(
     State(storage): State<Arc<T>>,
-    Json(pickup_message): Json<PickupMsgEnum>,
-) -> (StatusCode, Json<PickupMsgEnum>) {
+    Json(pickup_message): Json<Pickup>,
+    auth_pubkey: &str,
+) -> (StatusCode, Json<Pickup>) {
     match &pickup_message {
-        PickupMsgEnum::PickupStatusReqMsg(status_request) => (
+        Pickup::StatusRequest(status_request) => (
             StatusCode::OK,
-            handle_pickup_status_req(status_request, storage).await,
+            handle_pickup_status_req(&status_request.content, storage, auth_pubkey).await,
         ),
         // Why is client sending us status? That's server's job.
-        PickupMsgEnum::PickupStatusMsg(_status) => (
+        Pickup::Status(_status) => (
             StatusCode::BAD_REQUEST,
-            handle_pickup_type_not_implemented().await,
+            handle_pickup_default_status(storage, auth_pubkey).await,
```

Comment:
With respect to your comment above, what is this handling we call here? We should probably really just return `BAD_REQUEST` http response, but I don't see what any `handle_*` function is actually called upon receiving invalid request.

On slightly other note, from mediator client perspective, even more graceful response than getting `400` HTTP error, would be to actually return `200` with properly encrypted `ProblemReport` reporting that a message they submitted was unexpected.

Reply:
Currently, I am responding with a simple Pickup Status message on receipt of any non-request type pickup message that the client sends.

Here for example, if client sends mediator a `status` message (they should be sending `status-request`), then we respond with the correct `status` message, assuming that is what the client actually, wants. 

Agreed, problem report feels more elegant. For the time being, I was having the handler functions map from Pickup (receipt) to Pickup (response). To keep everything within the same protocol at protocol handler level. 

Ideally, these protocol handler functions should return a Result<Pickup> so the error case can be converted to Problem Report messages. In combination with your `MediatorError` suggestion, for example. 

Is it okay if I do this as part of another PR, as it'd be a change for all protocols and not just pickup.

Reply:
> Here for example, if client sends mediator a status message (they should be sending status-request), then we respond with the correct status message, assuming that is what the client actually, wants.

You are too kind to these buggy clients 😆 

> do this as part of another PR
Fair enough, let's leave this for later